src/L1/rollup/ScrollChain.sol

```diff
@@ -583,10 +623,83 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
         emit FinalizeBatch(_batchIndex, _batchHash, _postStateRoot, _withdrawRoot);
     }
 
+    /// @inheritdoc IScrollChain
+    function finalizeBundleWithTeeProof(
+        bytes calldata _batchHeader,
+        bytes32 _postStateRoot,
+        bytes32 _withdrawRoot,
+        bytes calldata _teeProof
+    ) external override OnlyProver whenFinalizeNotPaused {
+        // compute pending batch hash and verify
+        (uint256 batchPtr, bytes32 _batchHash, uint256 _batchIndex, ) = _loadBatchHeader(_batchHeader);
+
+        // check if this batch is finalized by zk proof.
```

Comment:
```suggestion
        // check if this state root is verified by zk proof.
```

Just a nitpick. It's still possible that the batch is "finalized" by zkp but it's in the middle of the bundle, so we do not have it stored in state.